sys/thread.c

```diff
@@ -131,16 +136,18 @@ noreturn void thread_exit(int exitcode) {
   fdtab_release(td->td_fdtable);
 
   mtx_lock(&zombie_threads_mtx);
+  TAILQ_INSERT_TAIL(&zombie_threads, td, td_zombieq);
+  mtx_unlock(&zombie_threads_mtx);
 
   critical_enter();
```

Comment:
Seems like we don't need `critical_enter` here as long as scheduler respects `td_lock` mutex. Please have a look at preferred solution [kern_thread.c:488](http://bxr.su/FreeBSD/sys/kern/kern_thread.c#488). You'll need to add `sched_abandon` function that removes a thread from scheduler and makes sure it will not ever be returned to.